### ISSUE TITLE
WHEN Trigger for Things

### ISSUE THREAD

**Issue by mhbosch** [openhab-scripters/openhab-helper-libraries#379]
Hello,

it is possible to add a When Trigger for Things? 
With the DSL Rule, i can observe a thing, with Jython i dont have a trigger for things.

My example for DSL:
```
rule "Gerät gestört"
when 
 Thing 'homematic:HM-LC-Sw4-DR:PEQ0172029:KEQ0055288' changed from ONLINE to OFFLINE
 then
 logInfo("Heizung", "Die Heizungskontrolle kann nicht erreicht werden")
   sendTelegram("bot1", "Die Heizungskontrolle kann nicht erreicht werden")
end

rule "Gerät gestört"
when 
 Thing 'homematic:HM-LC-Sw4-DR:PEQ0172029:KEQ0055288' changed from OFFLINE to ONLINE
 then
 logInfo("Heizung", "Die Heizungskontrolle kann wieder erreicht werden")
   sendTelegram("bot1", "Die Heizungskontrolle kann wieder erreicht werden")
end

```

**Comment by CrazyIvan359** [comment]
It is possible, there is an example in the docstring for `when`:

https://github.com/openhab-scripters/openhab-helper-libraries/blob/03038de498f70e6fa46d2a1457c38c8a4902c4e5/Core/automation/lib/python/core/triggers.py#L62

**Comment by mhbosch** [comment]
Hello,

thanks. I have seen, the string for the Thing was changed with OH3.
Now i tried it a sceond time with the other String and it will works!

**Comment by mhbosch** [comment]
Hello,

im again.

The Trigger dont have a action.

My Trigger:
```
@rule("Heizungskontrolle", description="Aenderung bei der Heizungskontrolle", tags=["Script", "Telegram"])
@when("Thing homematic:HM-LC-Sw4-DR:3014F711A061A7D8A98C07FD:KEQ0055288 changed")
def heizungskontrolle(event):
    LogAction.logInfo(u"Heiungskontrolle", "" + ir.getItem(str(event.itemName)).label + " wurde von " + str(event.oldItemState) + " auf " + str(event.itemState) + "geändert")
    actions.get("telegram", "telegram:telegramBot:Telegram_Bot").sendTelegram( ir.getItem(str(event.itemName)).label + " wurde von " + str(event.oldItemState) + " auf " + str(event.itemState) + "geändert")
```

The Log shows, the Trigger was accepted:
`2021-01-08 11:09:45.109 [DEBUG] [jsr223.jython.core.triggers         ] - when: target: 'Thing homematic:HM-LC-Sw4-DR:3014F711A061A7D8A98C07FD:KEQ0055288 changed', target_type: 'Thing', trigger_target: 'homematic:HM-LC-Sw4-DR:3014F711A061A7D8A98C07FD:KEQ0055288', trigger_type: 'changed', old_state: 'None', new_state: 'None'`

When i have a change, the trigger dont work:
LOG:
`Thing 'homematic:HM-LC-Sw4-DR:3014F711A061A7D8A98C07FD:KEQ0055288' changed from OFFLINE (COMMUNICATION_ERROR) to ONLINE`

But no error from script or active trigger.

**Comment by CrazyIvan359** [comment]
I have finally had some time to work on OH and was working near this trigger, so I dug into it. I thought it was using the `ThingStatusChangeTrigger`, which is built in to openHAB, but it only uses that if you specify a state. Turns out it is using a `GenericEventTrigger` and those have been broken for a while, but the good news is I have fixed them!

I am working on several fixes that I will be making available in the [ivans-updates](https://github.com/CrazyIvan359/openhab-helper-libraries/tree/ivans-updates) branch of my fork until such time as I can get them merged here, the fix for your issue will be the first.

**Comment by mhbosch** [comment]
Hello,

thats sound great. It there a Update Script for the library? I think, i miss updates.

**Comment by CrazyIvan359** [comment]
No, there isn't one. Did you use the manual installation instructions and are you running OH3?

**Comment by mhbosch** [comment]
Yes, OH3.

I hope, in near future, there is a script, who inform about updates and a update script,

**Comment by CrazyIvan359** [comment]
Might just be easier to modify the code in your file for now. Open `{OH_CONF}/automation/lib/python/core/triggers.py` and scroll way down to `class ThingEventTrigger(Trigger):` and replace the existing `__init__` function with this one:

```python
    def __init__(self, event_types, thing_uid=None, trigger_name=None):
        trigger_name = validate_uid(trigger_name)
        self.trigger = TriggerBuilder.create().withId(trigger_name).withTypeUID("core.GenericEventTrigger").withConfiguration(Configuration({
            "eventTopic": "*/things/*",
            "eventSource": "/things/{}".format(thing_uid) if thing_uid else "",
            "eventTypes": event_types
        })).build()
```

**Comment by mhbosch** [comment]
OK,

thanks, i will try it.
